wgpu-core/src/device/life.rs

```diff
@@ -14,111 +14,81 @@ use crate::{
+    pub(crate) bind_groups: FastHashMap<BindGroupId, Arc<BindGroup<A>>>,
+    pub(crate) bind_group_layouts: FastHashMap<BindGroupLayoutId, Arc<BindGroupLayout<A>>>,
+    pub(crate) render_pipelines: FastHashMap<RenderPipelineId, Arc<RenderPipeline<A>>>,
+    pub(crate) compute_pipelines: FastHashMap<ComputePipelineId, Arc<ComputePipeline<A>>>,
+    pub(crate) pipeline_layouts: FastHashMap<PipelineLayoutId, Arc<PipelineLayout<A>>>,
+    pub(crate) render_bundles: FastHashMap<RenderBundleId, Arc<RenderBundle<A>>>,
+    pub(crate) query_sets: FastHashMap<QuerySetId, Arc<QuerySet<A>>>,
 }
 
-impl ResourceMaps {
-    fn add_type<Id, R>(&mut self) -> &mut Self
-    where
-        Id: id::TypedId,
-        R: Resource<Id>,
-    {
-        let map = FastHashMap::<Id, Arc<R>>::default();
-        self.maps.insert(R::TYPE, Box::new(map));
-        self
-    }
-    fn map_mut<Id, R>(&mut self) -> &mut FastHashMap<Id, Arc<R>>
-    where
-        Id: id::TypedId,
-        R: Resource<Id>,
-    {
-        let map = self
-            .maps
-            .entry(R::TYPE)
-            .or_insert_with(|| Box::<FastHashMap<Id, Arc<R>>>::default());
-        let any_map = map.as_mut().as_any_mut();
-        let map = any_map.downcast_mut::<FastHashMap<Id, Arc<R>>>().unwrap();
-        map
-    }
-    pub(crate) fn new<A: HalApi>() -> Self {
-        let mut maps = Self::default();
-        maps.add_type::<BufferId, Buffer<A>>();
-        maps.add_type::<StagingBufferId, StagingBuffer<A>>();
-        maps.add_type::<TextureId, Texture<A>>();
-        maps.add_type::<TextureViewId, TextureView<A>>();
-        maps.add_type::<SamplerId, Sampler<A>>();
-        maps.add_type::<BindGroupId, BindGroup<A>>();
-        maps.add_type::<BindGroupLayoutId, BindGroupLayout<A>>();
-        maps.add_type::<RenderPipelineId, RenderPipeline<A>>();
-        maps.add_type::<ComputePipelineId, ComputePipeline<A>>();
-        maps.add_type::<PipelineLayoutId, PipelineLayout<A>>();
-        maps.add_type::<RenderBundleId, RenderBundle<A>>();
-        maps.add_type::<QuerySetId, QuerySet<A>>();
-        maps
+impl<A: HalApi> ResourceMaps<A> {
+    pub(crate) fn new() -> Self {
+        ResourceMaps {
+            buffers: FastHashMap::default(),
+            staging_buffers: FastHashMap::default(),
+            textures: FastHashMap::default(),
+            texture_views: FastHashMap::default(),
+            samplers: FastHashMap::default(),
+            bind_groups: FastHashMap::default(),
+            bind_group_layouts: FastHashMap::default(),
+            render_pipelines: FastHashMap::default(),
+            compute_pipelines: FastHashMap::default(),
+            pipeline_layouts: FastHashMap::default(),
+            render_bundles: FastHashMap::default(),
+            query_sets: FastHashMap::default(),
+        }
     }
+
     pub(crate) fn clear(&mut self) {
-        self.maps.iter_mut().for_each(|(_t, map)| map.clear_map());
+        self.buffers.clear();
```

Comment:
It's longer, but both here and in `extend` it would be way better to use destructuring on the struct. This way there's a guarantee that none of the fields is missed. This makes this safer against resource types being added and as as a reviewer I refuse to manually check completelnes given that there's a compiler mechanism to do so that will warn when one of the destructured elements is unused ;)

i.e. this
``` 
let ResourceMaps {
    staging_buffers,
    textures,
    // ...
} = &mut self;
staging_buffers.clear()
// ...
```

Reply:
The dreadful `clippy::pattern_type_mismatch` lint really gets in the way of that, sadly.

Reply:
I disabled the lint locally and will petition to get rid of it globally in a followup.